Networking/Networking/ApplicationPassword/RequestProcessor.swift

```diff
@@ -0,0 +1,87 @@
+import Alamofire
+import Foundation
+
+/// Authenticates and retries requests
+///
+final class RequestProcessor {
+    private var requestsToRetry = [RequestRetryCompletion]()
+
+    private var isAuthenticating = false
+
+    private let requestAuthenticator: RequestAuthenticator
+
+    init(credentials: Credentials?) {
+        requestAuthenticator = RequestAuthenticator(credentials: credentials)
+    }
+}
+
+// MARK: Request Authentication
+//
+extension RequestProcessor: RequestAdapter {
+    func adapt(_ urlRequest: URLRequest) throws -> URLRequest {
+        return try requestAuthenticator.authenticate(urlRequest)
+    }
+}
+
+// MARK: Retrying Request
+//
+extension RequestProcessor: RequestRetrier {
+    func should(_ manager: Alamofire.SessionManager,
+                retry request: Alamofire.Request,
+                with error: Error,
+                completion: @escaping Alamofire.RequestRetryCompletion) {
+        guard
+            request.retryCount == 0, // Only retry once
+            let urlRequest = request.request,
+            requestAuthenticator.shouldRetry(urlRequest), // Retry only REST API requests that use application password
+            shouldRetry(error) // Retry only specific errors
+        else {
+            return completion(false, 0.0)
+        }
+
+        requestsToRetry.append(completion)
+        if !isAuthenticating {
+            generateApplicationPassword()
+        }
+    }
+}
+
+// MARK: Helpers
+//
+private extension RequestProcessor {
+    func generateApplicationPassword() {
+        Task(priority: .medium) {
+            isAuthenticating = true
+
+            do {
+                let _ = try await requestAuthenticator.generateApplicationPassword()
+                isAuthenticating = false
+                completeRequests(true)
+            } catch {
+                isAuthenticating = false
+                completeRequests(false)
+            }
```

Comment:
just to clarify: on completion, it would trigger `adapt(_:)` again with an application password if `retry == true` right?

Reply:
Yes, the `adapt(_:)` method will be called for the retried requests. (which will authenticate the requests)`completeRequests` [method](https://github.com/woocommerce/woocommerce-ios/pull/8483/files#diff-e2b63cf38f66c7a410720d0644c041faf8dbbca67d1e011e7ce325bd698a596dR81-R86) will retry all the requests from the `requestsToRetry` array if the application password generation is successful.